EventFilter/DTRawToDigi/plugins/DTuROSRawToDigi.cc

```diff
@@ -0,0 +1,466 @@
+  produces<std::vector<DTuROSFEDData>>();
+
+  DTuROSInputTag_ = pset.getParameter<edm::InputTag>("inputLabel");
+
+  debug_ = pset.getUntrackedParameter<bool>("debug", false);
+
+  feds_ = pset.getUntrackedParameter<std::vector<int> >("feds", std::vector<int>());
+ 
+  nfeds_ = feds_.size();
+
+  Raw_token = consumes<FEDRawDataCollection>(DTuROSInputTag_);
+
+}
+
+
+DTuROSRawToDigi::~DTuROSRawToDigi(){}
+
+
+void DTuROSRawToDigi::produce(edm::Event& e, const edm::EventSetup& c) {
+
+  DTDigiCollection digis;
+  std::vector<DTuROSFEDData> words;
+
+  if (!fillRawData(e, c, digis, words)) return;
+
+  auto uROSDTDigi_product = std::make_unique<DTDigiCollection>(digis);
+  auto uROSDTWord_product = std::make_unique<std::vector<DTuROSFEDData>>(words);
+
+  e.put(std::move(uROSDTDigi_product));
+  e.put(std::move(uROSDTWord_product));
+
+}
+
+
+bool DTuROSRawToDigi::fillRawData(edm::Event& e, const edm::EventSetup& c,
+                                  DTDigiCollection& digis, std::vector<DTuROSFEDData>& words) {
+
+  edm::Handle<FEDRawDataCollection> data;
+  e.getByToken(Raw_token, data);
+
+  edm::ESHandle<DTReadOutMapping> mapping;
+  c.get<DTReadOutMappingRcd>().get( mapping );
+
+  for (int w_i = 0; w_i < nfeds_; ++w_i) {
+    DTuROSFEDData fwords;
+    process(feds_[w_i], data, mapping, digis, fwords);
+    words.push_back(fwords);
+  }
+  
+  return true;
+}
+
+
+void DTuROSRawToDigi::process(int DTuROSFED,
+                              edm::Handle<FEDRawDataCollection> data,
+                              edm::ESHandle<DTReadOutMapping> mapping,
+                              DTDigiCollection& digis,
+                              DTuROSFEDData& fwords) {
+
+
+  // Container
+  std::vector<long> DTuROSWordContainer;
+
+
+  // Header constituents
+  int BOEevTy, DTuROSId;
+
+
+  // Trailer constituents
+  int chkEOE, evtLgth, CRC;
+
+
+  // Hit counter
+  std::map<uint32_t, int> hitOrder;
+
+
+  //--> Header - line 1
+
+
+  FEDRawData dturosdata = data->FEDData(DTuROSFED); 
+  if ( dturosdata.size() == 0 ) return;
+
+
+  lineFED=dturosdata.data(); 
+  long dataWord = 0;
+  int lines  = 0; // counting including header
+  readline(lines, dataWord);
+
+
+  BOEevTy  = ( dataWord >> 60 ) & 0xF;  // positions 60 -> 63
+  DTuROSId = ( dataWord >> 8 ) & 0xFFF; // positions 8 -> 19
+
+
+  if ( (BOEevTy != 0x5) || (DTuROSId != DTuROSFED) ) {
+    if ( debug_ ) edm::LogWarning("dturos_unpacker") << "Not a DTuROS FED " << DTuROSFED << " or header " << std::hex << dataWord;
+    return;
+  }
+
+
+  fwords.setfed(DTuROSId);
+  fwords.setheader1(dataWord);
+
+
+  int newCRC = 0xFFFF;
+  dt_crc::calcCRC(dataWord, newCRC);
+
+  int crate = DTuROSId;
+
+
+  //--> Header - line 2
+
+
+  readline(lines, dataWord);
+  dt_crc::calcCRC(dataWord, newCRC);
+
+  int nslots = ( dataWord >> 52 ) & 0xF; // positions 52 -> 55
+
+
+  fwords.setheader2(dataWord);
+  fwords.setnslots(nslots);
+
+
+  //--> AMC - line 3 to 2+nslots
+  std::map<int, int> slot_size;
```

Comment:
this can be std::vector<int>(12) instead?

Reply:
No, it is the number of slots sending data in this event, hence it is not guaranteed to be 12 always.
Thanks

Reply:
indeed, I guess I meant std::vector<int>(nslots) - but right, the answer is no because the index is not j but slot when its filled.Thanks